File: 46-permutations/46-permutations.cpp
class Solution {
    void permute(vector<int>& nums, int i, vector<vector<int>>& result)
    {
        if(nums.size()<=i)
            result.push_back(nums);
        for(int begin = i ; begin < nums.size() ; begin++)
        {
            swap(nums[i],nums[begin]);
            permute(nums,i+1,result);
            swap(nums[i],nums[begin]);
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        permute(nums,0,result);
        return result;
    }
};

File: 46-permutations/46-permutations.java
class Solution {
    List<List<Integer>> list = new ArrayList(new ArrayList());
    public List<List<Integer>> permute(int[] nums) {
       
        generatePermutations(nums,0,nums.length-1);
        return list;
    }
    
    void generatePermutations(int nums[],  int i, int n) {
        if(i == n) {
            List<Integer> l = IntStream.of(nums).boxed().collect(Collectors.toList());
            list.add(l);
            return;
        }
        for (int j = i;j<=n;j++) {
            swap(nums, i, j);
            generatePermutations(nums, i+1,n);
            swap(nums,i,j);
        }
    }
    
    void swap(int nums[], int i , int j) {
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }
}

// 0 2